src/os/bluestore/BlueFS.cc

```diff
@@ -1805,6 +1805,15 @@ int BlueFS::_allocate(uint8_t id, uint64_t len,
   int r = -ENOSPC;
   if (alloc[id]) {
     r = alloc[id]->reserve(left);
+    if (r < 0) {
+      vector<interval_set<uint64_t>> to_release(pending_release.size());
+      to_release.swap(pending_release);
+      for (unsigned i = 0; i < to_release.size(); ++i) {
+	for (auto p = to_release[i].begin(); p != to_release[i].end(); ++p) {
+	  alloc[i]->release(p.get_start(), p.get_len());
+	}
+      }
+    }
```

Comment:
we need to retry the reserve() call, then?

Reply:
Actualy, no, this still isn't right.. we have to flush the log before we can release and potentially re-use (i.e. write to) previously referenced extents or else we may crash and have overwritten data we previously referenced. 

If we want to improve this then we should be able to move that release dance in sync_metadata() into _sync_and_flush_log().  I.e., we need to repvent a case where an unlink causes us to release extents, reallocate, and overwrite old data before the unlink event has committed to the log..

Reply:
Ok. But  i have a question: your commit :e0ec06ed718e3b3197a9e53111bcdbf4e3190086 which introduce deferred release unused extents. I'm not understand. Why? Thanks

Reply:
unlink A: release extents(a,b,c...)
A allocate: a,b.
But unlink A will update log_t before log_t update when allocate(a,b). If crash, those logs all missed. Which case cause log(allocate) exist but log(release) missed.